src/function.rs

```diff
@@ -136,3 +141,42 @@ fn just_executable(_context: &FunctionContext) -> Result<String, String> {
     )
   })
 }
+
+fn file_name(_context: &FunctionContext, file_path: &str) -> Result<String, String> {
+  Utf8Path::new(file_path)
+    .file_name()
+    .map(std::borrow::ToOwned::to_owned)
+    .ok_or_else(|| format!("Cannot get file name from `{}`", file_path))
+}
+
+fn parent_directory(_context: &FunctionContext, file_path: &str) -> Result<String, String> {
+  Utf8Path::new(file_path)
+    .parent()
+    .map(Utf8Path::as_str)
+    .map(std::borrow::ToOwned::to_owned)
+    .ok_or_else(|| format!("Cannot get parent directory from `{}`", file_path))
+}
+
+fn file_stem(_context: &FunctionContext, file_path: &str) -> Result<String, String> {
+  Utf8Path::new(file_path)
+    .file_stem()
+    .map(std::borrow::ToOwned::to_owned)
+    .ok_or_else(|| format!("Cannot get file stem from `{}`", file_path))
+}
+
+fn without_extension(_context: &FunctionContext, file_path: &str) -> Result<String, String> {
+  let path = Utf8Path::new(file_path);
```

Comment:
I think it would be better to do this in two steps, first get the parent and report an error if you can't get it, then get the file_stem and report an error, and then return the error.

Also, you can do `Utf8Path::join` instead of `format!("{}/{}", …)`.

Reply:
For example:

```
 let parent = Utf8Path::new(file_path).parent().ok_or_else(|| format!("Cannot extract parent from `{}`, path)?;
…
```